MRG: concatenate_raws(), concatenate_epochs(), write_evokeds() gain on_mismatch param and raise by default

hoechenberger opened the pull request:
`mne.concatenate_raws()` gains on_mismatch param and raises by default

Fixes #9436

hoechenberger commented:
@larsoner I put the changelog entry into the API section, but it could just as well go to Bugs. WDYT?

agramfort commented on mne/io/base.py:
```diff
@@ -2422,6 +2428,28 @@ def concatenate_raws(raws, preload=None, events_list=None, verbose=None):
+                  allowed_values=('raise', 'warn', 'ignore'))
+
+    first_dev_head_t = raws[0].info['dev_head_t']
+    for raw in raws[1:]:
```
Can you factorize the code with concatenate_epochs ?

hoechenberger replied:
@agramfort I looked into this but we're going so much epochs-specific magic in `concatenate_epochs()` that I felt this would actually complicate things

hoechenberger replied:
No, wait. I think I can (re)use `_compare_epochs_infos()`. Need to look into this.

larsoner replied:
I expect there to be some pain involved (maybe a new param for `concatenate_epochs`, maybe another BUG entry, etc.), but agree with @agramfort that at least the info-checking functionality should be shared between the two. I don't in principle see why the info comparisons should differ. Let's see if you agree from looking at the code, I don't know how they are different...

hoechenberger replied:
I've tried to address your comments in b087b5b

larsoner commented on mne/evoked.py:
```diff
@@ -1354,7 +1354,7 @@ def write_evokeds(fname, evoked):
 def _write_evokeds(fname, evoked, check=True):
     """Write evoked data."""
-    from .epochs import _compare_epochs_infos
+    from .io.meas_info import _ensure_infos_match
```
this does not need to be nested anymore and can be combined with the existing `from .io.meas_info` line at the top

larsoner commented on mne/epochs.py:
```diff
@@ -3366,7 +3334,7 @@ def _concatenate_epochs(epochs_list, with_data=True, add_offset=True):
     events_offset = _update_offset(None, out.events, shift)
     for ii, epochs in enumerate(epochs_list[1:], 1):
-        _compare_epochs_infos(epochs.info, info, ii)
+        _ensure_infos_match(epochs.info, info, ii)
```
```suggestion
        _ensure_infos_match(epochs.info, info, f'epochs[{ii}]')
```

larsoner commented on mne/io/meas_info.py:
```diff
@@ -2478,3 +2478,60 @@ def _write_ch_infos(fid, chs, reset_range, ch_names_mapping):
+    _check_option(parameter='on_mismatch', value=on_mismatch,
+                  allowed_values=('raise', 'warn', 'ignore'))
+
+    if not isinstance(name, str):  # passed epochs index
+        name = f'epochs[{name:d}]'
```
```suggestion
```

larsoner commented on mne/io/meas_info.py:
```diff
@@ -2478,3 +2478,60 @@ def _write_ch_infos(fid, chs, reset_range, ch_names_mapping):
+    ----------
+    info1, info2 : instance of Info
+        The infos to compare.
+    name : str or int
```
```suggestion
    name : str
```

larsoner commented on mne/io/meas_info.py:
```diff
@@ -2478,3 +2478,60 @@ def _write_ch_infos(fid, chs, reset_range, ch_names_mapping):
+    info2._check_consistency()
+
+    msg = None
+    if info1['nchan'] != info2['nchan']:
+        msg = f"{name}.info['nchan'] values don't match"
+    elif set(info1['bads']) != set(info2['bads']):
+        msg = f"{name}.info['bads'] don't match"
+    elif info1['sfreq'] != info2['sfreq']:
+        msg = f"{name}.info['sfreq'] don't match"
+    elif set(info1['ch_names']) != set(info2['ch_names']):
+        msg = f"{name}.info['ch_names'] don't match"
+    elif len(info2['projs']) != len(info1['projs']):
+        msg = f"SSP projectors in {name} don't match"
```
I don't think we want to do this `msg` for *all* options, just for `info['dev_head_t']`. The others should just directly raise an exception as they did before

larsoner commented on mne/io/base.py:
```diff
@@ -2409,10 +2410,15 @@ def concatenate_raws(raws, preload=None, events_list=None, verbose=None):
+        How to handle situations when we discover that specific aspects of the
+        `~mne.io.Raw` instances mismatch in a way that could cause problems
+        after concatenation. Currently, this check is limited to the
+        device-to-head transformation. Defaults to ``'raise'``.
```
What you say here about `device-to-head`-only is how it should work, but this is not how it's coded

larsoner commented on mne/io/meas_info.py:
```diff
@@ -2478,3 +2478,60 @@ def _write_ch_infos(fid, chs, reset_range, ch_names_mapping):
+               f"runs to a common head position.")
+
+    if msg is not None:
+        if on_mismatch == 'raise':
+            raise ValueError(msg)
+        elif on_mismatch == 'warn':
+            warn(msg)
+        else:
+            logger.info(msg)
```
Should use `_on_missing`, `git grep on_missing` to see how it's used elsewhere

larsoner commented on mne/io/meas_info.py:
```diff
@@ -2478,3 +2478,60 @@ def _write_ch_infos(fid, chs, reset_range, ch_names_mapping):
+    on_mismatch : 'raise' | 'warn' | 'ignore'
+        What to do in case of a mismatch between `info1` and `info2`.
+    """
+    _check_option(parameter='on_mismatch', value=on_mismatch,
+                  allowed_values=('raise', 'warn', 'ignore'))
```
Should use `_check_on_missing`

hoechenberger commented:
@larsoner I believe I've addressed all of your comments

larsoner commented on mne/io/base.py:
```diff
@@ -2409,10 +2410,13 @@ def concatenate_raws(raws, preload=None, events_list=None, verbose=None):
     events_list : None | list
-        The events to concatenate. Defaults to None.
+        The events to concatenate. Defaults to ``None``.
+    on_mismatch : 'raise' | 'warn' | 'ignore'
+        How to handle situations where the device-to-head transformation
+        differs between raw instances. Defaults to ``'raise'``.
```
Should make use of `_on_missing_base` and docdict, see `mne/utils/docs.py`

Also needs `.. versionadded` directive

hoechenberger replied:
Argh... yes to all. Thanks for taking your time

larsoner commented on mne/epochs.py:
```diff
@@ -3366,7 +3334,7 @@ def _concatenate_epochs(epochs_list, with_data=True, add_offset=True):
     events_offset = _update_offset(None, out.events, shift)
     for ii, epochs in enumerate(epochs_list[1:], 1):
-        _compare_epochs_infos(epochs.info, info, ii)
+        _ensure_infos_match(epochs.info, info, f'epochs[{ii}]')
```
I think you should also add `on_mismatch` to `concatenate_epochs`, since it's very similar to raw. I guess it should also be done for evoked?

hoechenberger replied:
Ok!

hoechenberger commented:
@larsoner Ok next round :)

I've added the `on_mismatch` param to more public functions, as you proposed. I've added it as kw-only argument, which I **thought** could make sense here, but happy to revert this if you feel differently.

hoechenberger commented:
@larsoner Ok. I hope that's it. 😅👌

hoechenberger commented:
The failing CI job actually ran all tests successfully. That is, all CI tests pass

agramfort commented:
Not sure if the CI failure is legit. Otherwise lgtm. @larsoner feel free to merge if happy

larsoner commented on doc/changes/latest.inc:
```diff
@@ -63,4 +66,4 @@ Bugs
 API changes
 ~~~~~~~~~~~
-- Nothing yet
+- :func:`mne.concatenate_raws` now raises an exception if ``raw.info['dev_to_head_t']`` differs between files. This behavior can be controlled using the new ``on_mismatch`` parameter (:gh:`9438` by `Richard Höchenberger`_)
```
This is an immediate change in behavior without a deprecation cycle, so it has to live in the BUG section

larsoner commented on mne/utils/docs.py:
```diff
@@ -58,6 +58,13 @@
 """ % (_on_missing_base,)
+docdict['on_info_mismatch'] = f"""
+on_mismatch : 'raise' | 'warn' | 'ignore'
+    {_on_missing_base} the device-to-head transformation differs between raw
```
Not just raw anymore
```suggestion
    {_on_missing_base} the device-to-head transformation differs between
```

larsoner commented:
Other than two minor remaining gripes, LGTM +1 for merge

hoechenberger commented:
All done here. 😁😅

hoechenberger commented:
Circle was failing because of some HTTP timeouts. I restarted the Circle build.

agramfort commented:
Thx 🙏 @hoechenberger
